grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -1089,6 +1094,106 @@ class StockMovementService {
+                    Boolean lotMatches = params.lot == item.inventoryItem?.lotNumber
+                    binLocationMatches && lotMatches
+                }
+
+                if (!availableItem) {
+                    params.errors.push("There is no item available with: ${params.lot ? "lot ${params.lot}" : ""} ${params.binLocation ? "bin ${params.binLocation}" : ""}")
+                }
+            }
+        }
+
+        Boolean isUnderPickAllowed = Holders.config.openboxes.picklist.underpick.enabled
+        Boolean isOverPickAllowed = Holders.config.openboxes.picklist.overpick.enabled
+
+        data.groupBy { it.id }.each { itemId, params ->
+            PickPageItem pickPageItem = pickPageItems.find { it.requisitionItem?.id == itemId }
+            Integer itemQuantitySum = params.sum { Integer.parseInt(it.quantity) }
+            if (!isOverPickAllowed && pickPageItem && itemQuantitySum > pickPageItem.requisitionItem.quantity) {
+                errors.push("Item ${itemId} is overpicked, expedted quantity ${pickPageItem.requisitionItem.quantity}, received ${itemsSum}")
+            }
+            if (!isUnderPickAllowed && pickPageItem && itemQuantitySum < pickPageItem.requisitionItem.quantity) {
+                errors.push("Item ${itemId} is underpicked, expedted quantity ${pickPageItem.requisitionItem.quantity}, received ${itemsSum}")
+            }
+        }
+    }
+
+    void importPicklistTemplate(List<Map> data, StockMovement stockMovement, List<PickPageItem> pickPageItems) {
+        data.each { params ->
+            // skip rows with errors
+            if (params.errors) {
+                return
+            }
+
+            PickPageItem pickPageItem = pickPageItems.find {
+                it.requisitionItem?.id == params.id
+            }
+
+            AvailableItem availableItem = pickPageItem.availableItems?.find { item ->
+                Boolean binLocationMatches = params.binLocation ? item.binLocation?.name == params.binLocation : !item.binLocation
+                Boolean lotMatches = params.lot == item.inventoryItem?.lotNumber
+                binLocationMatches && lotMatches
+            }
+
+            RequisitionItem requisitionItem = pickPageItem.requisitionItem?.modificationItem ?: pickPageItem.requisitionItem
+
+            pickPageItem.picklistItems.each {
+                if (it.id) {
+                    it.quantity = 0
+                }
+            }
+            pickPageItem.picklistItems.add(new PicklistItem(
+                    requisitionItem: requisitionItem,
+                    inventoryItem: availableItem.inventoryItem,
+                    binLocation: availableItem.binLocation,
+                    quantity: params.quantity,
+                    sortOrder: pickPageItem.sortOrder
+            ))
+        }
+        createOrUpdatePicklistItem(stockMovement, pickPageItems)
```

Comment:
I can see that in this method there is the same issue that I've recently fixed. Could you please add in this method the association between requisition and the picklist?
`requisition.picklist = picklist`

Reply:
Can we just pass in the picklist items, not the pick page items? Or does that method not exist? If it doesn't we should create it and then start to refactor other methods to use the proper API.

FWIW I don't think this method should exist. It doesn't make sense from an API perspective. It feels like we're packing the required data into unnecessary proxy objects. 
```
void createOrUpdatePicklistItem(StockMovement stockMovement, List<PickPageItem> pickPageItems) 
```
Something like this would be better
```
void createOrUpdatePicklistItems(StockMovement stockMovement, List<PicklistItem> picklistItems) 
```
This would be even better
```
void createOrUpdatePicklistItems(Picklist picklist, List<PicklistItem> picklistItems) 
```
But we're also in a situation where it probably makes sense to do the persistence in the import method.
```
void importPicklistItems(List<Map> data, StockMovement stockMovement, 
    List<PickPageItem> pickPageItems) {

    ...
    Picklist picklist = stockMovement.requisition.picklist
    data.each { row ->
        PicklistItem picklistItem = new PicklistItem(row...)
        picklist.addToPicklistItems(picklistItem)
    }
    picklist.save()
}
```

Reply:
That performance issue sounds like maybe something else was happening though. But I take your point which seems to be "let's do things explicitly" vs "relying on Hibernate magic" :mage_woman: 

But in my mind, the main reasons to use Hibernate is to 
* manage your associations (cascades)
* avoid costly saving (worse-case: flushing) each individual child object on its own. 
* provide a mechanism for rolling back a transaction

This is ignoring Hibernate altogether, could lead to performance degredation due to many and might not allow you to rollback the flushed data if an exception happens afterwards.
```
data.each { 
    PicklistItem picklistItem = new PicklistItem(it)
    picklistItem.picklist = picklist
    picklistItem.save(flush:true)
}
```

This approach let's Hibernate do its thing (based on cascading rules defined on the domain) and would rollback the picklist item data if there was a validation error in picklist or somewhere else. 
```
data.each { 
    PicklistItem picklistItem = new PicklistItem(it)
    picklist.addToPicklistItems(picklistItem)
}
picklist.save(flush:true)
```
This is all theoretical though. You seem to distrust Hibernate and I think that's a healthy attitude to have with respect to things that perform magic. 

So it would be a good discovery ticket to write this in multiple ways and then perform some integration tests to see which is the right approach. Then we can create a convention around it. 

* performance with a large number of picklist items
* can you rollback the data at different points in the Hiberate session/transaction
    * my hunch is that both would be rollback-able even if data has been saved with flush:true
* if we were to use different cascade behaviors (https://docs.grails.org/6.1.2/ref/Database%20Mapping/cascade.html), would either of these perform in a way that is not expected

Reply:
@jmiranda as for this one, I might be overcautious - I've relied on the cascade many times before and have not experienced any issues, but I also remember one situation (described above) that was due to the cascade saving.
I wouldn't say I want to do things explicitly, because I love Hibernate, and we should rely on it in most of the cases, e.g. I wouldn't say we should write:
```groovy
Picklist picklist = Picklist.get(...)
PicklistItem picklistItem = new PicklistItem(...)
picklistItem.picklist = picklist
picklist.picklistItems.add(picklistItem)
picklistItem.save()
```
instead of:
```groovy
Picklist picklist = Picklist.get(...)
PicklistItem picklistItem = new PicklistItem(...)
picklist.addToPicklistItems(picklistItem)
picklistItem.save()
```
because this would be an overkill. The recent bug we experienced around requisition/picklist was in 1:1 association where this have to be handled manually (because Hibernate adds it AFTER the session is closed, this is why it caused issues, whereas the `addTo*` (e.g. `addToPicklistItems` adds the association in two objects immediately))

An important note is that we hardly ever need to use `flush: true` while saving - I'd say the `flush: true` would be needed if for example constraints of the picklistItem rely on already existing picklistItems saved in the same transaction, to validate them properly.

In case of:
```groovy
data.each { 
  Picklist picklist = Picklist.get(...)
  PicklistItem picklistItem = new PicklistItem(...)
  picklist.addToPicklistItems(picklistItem)
  picklistItem.save()
}
```
Hibernate just creates a queue and triggers all the saves at the transaction commit.

So TL;DR would be - I trust in Hibernate magic and we should rely on it in most of the cases, but at the same time we need to know what happens behind the scenes and what might be the edge cases around some of the operations.
As for the performance issues around the cascade - I would skip it for now and check such things in a discovery/spike ticket around Hibernate - including the impact of the batchSize property, etc.